lib.py

```diff
@@ -0,0 +1,114 @@
+import asyncio
+import functools
+import traceback
+import os
+
+import docker
+from database import Database
+from messages import SubmitMessage
+
+doc = docker.from_env()
+async def build_image(msg, solution):
+    print(f"Building for {msg.author.name}")
+    status = await msg.reply("Building...")
+    with open("runner/src/code.rs", "wb+") as f:
+        f.write(solution)
+
+    loop = asyncio.get_event_loop()
+
+    try:
+        await loop.run_in_executor(None, functools.partial(doc.images.build, path="runner", tag=f"ferris-elf-{msg.author.id}"))
+        return True
+    except docker.errors.BuildError as err:
+        print("Build Error:")
+        traceback.print_exception(err)
+        e = ""
+        for chunk in err.build_log:
+            e += chunk.get("stream") or ""
+        await msg.reply(f"Error building benchmark: {err}")
+        return False
+    finally:
+        await status.delete()
+
+async def run_image(msg, input):
+    print(f"Running for {msg.author.name}")
+    # input = ','.join([str(int(x)) for x in input])
+    status = await msg.reply("Running benchmark...")
+    loop = asyncio.get_event_loop()
+    try:
+        out = await loop.run_in_executor(None, functools.partial(doc.containers.run, f"ferris-elf-{msg.author.id}", f"timeout 60 ./target/release/ferris-elf", environment=dict(INPUT=input), remove=True, stdout=True, mem_limit="24g", network_mode="none"))
+        out = out.decode("utf-8")
+        print(out)
+        return out
+    except docker.errors.ContainerError as err:
+        print("Run Error:")
+        traceback.print_exception(err)
+        await msg.reply(f"Error running benchmark: {err}")
+    finally:
+        await status.delete()
+
+async def benchmark(submit_msg: SubmitMessage):
+    (msg,day,code,part) = submit_msg.msg,submit_msg.day,submit_msg.code,submit_msg.part
+    build = await build_image(msg, code)
+    if not build:
+        return
+    
+    day_path = f"{day}/"
+    try:
+        onlyfiles = [f for f in os.listdir(day_path) if os.isfile(os.path.join(day_path, f))]
+    except:
```

Comment:
FIXME: This bare except should be replaced with the specific errors we want to catch, or at least be replaced with `except Exception as e`, since that wont stop BaseException's from being propagated (which are meant to be more similar to rust panics, BaseException's should not be caught or at least should be handled specially)

Reply:
fixed